### ISSUE TITLE
GELU produces incorrect results for negative ranges (goes to 0 faster than pytorch gelu)

### ISSUE THREAD

**Issue by skhorasganiTT** [tenstorrent-metal/tt-metal#5578]
GELU is producing incorrect values for small to large negative numbers. Specifically, it is going to 0 too fast (approximately when inputs become smaller than -2.5). This is degrading the PCC of the MLP in Falcon7B, and can likely affect other large models. 

This can be tested by modifying the eltwise gelu op test:

1. Modify `tests/tt_eager/python_api_testing/sweep_tests/pytests/tt_dnn/test_eltwise_unary.py::TestEltwiseUnary::test_run_eltwise_gelu_op` such that `datagen_func` has `high=-2` instead of `high=100`.

2. Run the test to observe the failure: 
```
pytest tests/tt_eager/python_api_testing/sweep_tests/pytests/tt_dnn/test_eltwise_unary.py::TestEltwiseUnary::test_run_eltwise_gelu_op[False-output_mem_config1-input_mem_config1-input_shapes0]
```

FYI @uaydonat

**Comment by bharane-abb** [comment]
> GELU is producing incorrect values for small to large negative numbers. Specifically, it is going to 0 too fast (approximately when inputs become smaller than -2.5). This is degrading the PCC of the MLP in Falcon7B, and can likely affect other large models.
> 
> This can be tested by modifying the eltwise gelu op test:
> 
> 1. Modify `tests/tt_eager/python_api_testing/sweep_tests/pytests/tt_dnn/test_eltwise_unary.py::TestEltwiseUnary::test_run_eltwise_gelu_op` such that `datagen_func` has `high=-2` instead of `high=100`.
> 2. Run the test to observe the failure:
> 
> ```
> pytest tests/tt_eager/python_api_testing/sweep_tests/pytests/tt_dnn/test_eltwise_unary.py::TestEltwiseUnary::test_run_eltwise_gelu_op[False-output_mem_config1-input_mem_config1-input_shapes0]
> ```
> 
> FYI @uaydonat

On debugging GELU, we observed an issue with tanh results. we have attached the document for reference, which includes step-by-step instructions on computing the GELU operation in both tt-metal and reference.
We also observed optimal input range for tanh(in TT implementation) is -2 to 2.

Document [link](https://docs.google.com/document/d/11L7Ak9JvF7MMtLgtKVtfg_OciqJjYhjbf2K-XS6T4k0/edit?usp=sharing)

**Comment by uaydonat** [comment]
fyi @rtawfik01

**Comment by rtawfik01** [comment]
After running a few experiments from my side, this is what I see:
1. Non-approx mode uses a cdf polynomial approximation function, approx mode uses 6 linear equations to approximate gelu regions.
2. Non-approx mode, while using inputs with random data in the region that Salar mentioned in the issue (-100 to -2), gives about ~0.9127 PCC, while approximate mode, gives about ~0.989PCC, with float16b
3. Using Float32 input/output still generates ~0.989 PCC for approx mode, but gives about ~0.938 PCC for non-approx mode in -100 to -2 range.
4. Non-approx mode starts giving zeroes for float16b at around x<=-2.5, while approx mode gives zeroes at around x<=-2.99
5. Approx mode starts producing positive values in the range -3 < x <=-2.999

So there are 2 potential issues:
- Gelu should not produce positive values with x < 0 (which occurs in -3 < x <=-2.999 for approx mode)
- Gelu is clamping to 0 for both approx & non-approx mode. For x <= -3, gelu is expected to get exponentially closer to 0. The question here is how much precision/range is expected for x <=-3 @skhorasganiTT? Do you expect non-zero results until x = -4 for example? What is the Falcon 7b dataset input to the gelu?

**Comment by skhorasganiTT** [comment]
@rtawfik01 @uaydonat I have added and tested evaluations on the AlpacaEval dataset for Falcon7b-prefill. These are the results I got for sequence length 256 with 32 sequences (did not run more sequences since the gelu-on-host experiment takes a relatively long time):

Default - gelu on device:
PCC: 0.970 | Top1 Acc: 0.896 | Top5 Acc: 0.999 | Max ATOL: 7.77

gelu on host:
PCC: 0.975 | Top1 Acc: 0.897 | Top5 Acc: 0.995 | Max ATOL: 6.88

gelu on device + fix [#5927](https://github.com/tenstorrent-metal/tt-metal/pull/5927):
PCC: 0.966 | Top1 Acc: 0.889 | Top5 Acc: 0.997 | Max ATOL: 8.08

no gelu in reference HF model or tt-metal model:
PCC: 0.999 | Top-1 Acc: 0.83 | Top5 Acc: 1.0 | Max ATOL: 1.88

Since the gelu-on-host experiment is giving similar results to the first experiment, the errors from gelu seem like they are not too significant. The main counter to this argument would be the last experiment which yields very high PCC without gelu, however that experiment is also showing a degradation in Top-1 accuracy, so it could be possible the PCC is misleading here.
